sched/sched/sched_note.c

```diff
@@ -631,144 +924,337 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
+ * Input Parameters:
+ *   oldm - A writable pointer to struct note_trace_mode_s to get current
+ *          trace mode
+ *          If 0, no data is written.
+ *   newm - A read-only pointer to struct note_trace_mode_s which holds the
+ *          new trace mode
+ *          If 0, the trace mode is not updated.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
 
-      note_remove();
+void sched_note_trace_mode(struct note_trace_mode_s *oldm,
+                           struct note_trace_mode_s *newm)
+{
+  irqstate_t irq_mask;
 
-      /* and return an error */
+  irq_mask = enter_critical_section();
 
-      notelen = -EFBIG;
-      goto errout_with_csection;
+  if (newm != NULL)
+    {
+      sched_note_trace_stop();
+      g_note_filter.mode = *newm;
+      if (g_sched_note_stream != NULL &&
+          g_sched_note_stream->bufclear != NULL)
+        {
+          (g_sched_note_stream)->bufclear(g_sched_note_stream,
+            (g_note_filter.mode.flag & TRIOC_MODE_FLAG_ONESHOT) != 0);
+        }
     }
 
-  /* Loop until the note has been transferred to the user buffer */
+  if (oldm != NULL)
+    {
+      *oldm = g_note_filter.mode;
+    }
 
-  remaining = (unsigned int)notelen;
-  while (remaining > 0)
+  leave_critical_section(irq_mask);
+}
+
+/****************************************************************************
+ * Name: sched_note_trace_syscallfilter
+ *
+ * Description:
+ *   Set and get syscall trace filter setting
+ *   (Same as TRIOC_GETSYSCALLFILTER / TRIOC_SETSYSCALLFILTER ioctls)
+ *
+ * Input Parameters:
+ *   oldf - A writable pointer to struct note_trace_syscallfilter_s to get
+ *          current syscall trace filter setting
+ *          If 0, no data is written.
+ *   newf - A read-only pointer to struct note_trace_syscallfilter_s of the
+ *          new syscall trace filter setting
+ *          If 0, the setting is not updated.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+void sched_note_trace_syscallfilter(struct note_trace_syscallfilter_s *oldf,
+                                    struct note_trace_syscallfilter_s *newf)
+{
+  irqstate_t irq_mask;
+  int i;
+  int nr;
+
+  irq_mask = enter_critical_section();
+
+  if (newf != NULL)
     {
-      /* Copy the next byte at the tail index */
+      sched_note_trace_stop();
 
-      *buffer++ = g_note_info.ni_buffer[tail];
+      /* Replace the syscall filter mask by the provided setting */
 
-      /* Adjust indices and counts */
+      memset(g_note_filter.syscall_mask, 0,
+             sizeof g_note_filter.syscall_mask);
+      for (i = 0; i < newf->nr_syscalls; i++)
+        {
+          nr = newf->syscalls[i];
 
-      tail = note_next(tail, 1);
-      remaining--;
+          if (nr < SYS_nsyscalls)
+            {
+              BITMASK_SET(g_note_filter.syscall_mask, nr);
```

Comment:
should we use the same format here to avoid the conversion(number<->bitmap)?

Reply:
Providing the interface like fd_set used by select() ? It seems good.